packages/mui-base/src/MenuItemUnstyled/MenuItemUnstyled.tsx

```diff
@@ -42,25 +43,25 @@ const MenuItemUnstyled = React.forwardRef(function MenuItemUnstyled(
     ...other
   } = props;
 
-  const Root = component ?? components.Root ?? 'li';
-
   const { getRootProps, disabled, focusVisible } = useMenuItem({
     disabled: disabledProp,
     ref,
     label,
   });
 
-  const ownerState: MenuItemOwnerState = { ...props, disabled, focusVisible };
+  const ownerState: MenuItemUnstyledOwnerState = { ...props, disabled, focusVisible };
 
   const classes = getUtilityClasses(ownerState);
 
+  const Root = component ?? components.Root ?? 'li';
+  const rootComponentProps = resolveComponentProps(componentsProps.root, ownerState);
   const rootProps = appendOwnerState(
     Root,
     {
       ...other,
-      ...componentsProps.root,
       ...getRootProps(other),
-      className: clsx(classes.root, className, componentsProps.root?.className),
+      ...rootComponentProps,
```

Comment:
This way devs are able to override whatever the hook is setting. But I see there's a problem with event handlers. I suppose the order should be as following:

1. Props from the hook (with event handlers passed in to be chained)
2. Additional props from the component (`...other`) with removed event handlers - only for the root slot
3. `componentsProps.<slot name>` with removed event handlers
4. `className` - merged from `className` prop, `componentsProps.*.className` and slot-specific classes
5. `ref` - if necessary

Developers would be able to override whatever prop they need and if they provide an event handler, it would be chained with the hook's handlers.

We could use a function that would encapsulate this logic and ensure this works consistently across our components.

What do you think?

Reply:
Yep, let's do this 👍